### DISCUSSION TITLE
SetDisplayCallbacks

### DISCUSSION THREAD

**Discussion by maniglia** [K-Society/KSociety.SharpCubeProgrammer#89] [Q&A]
The _SetDisplayCallbacks_ function is used in _Programmer.dll_ [(dllmain)](https://github.com/K-Society/KSociety.SharpCubeProgrammer/blob/master/src/01/Programmer/dllmain.cpp) and disable the functionality of logging any message generated by the _CubeProgrammer_API.dll_ library to maximize performance.

**Comment by mihaisvcc** [comment]
Hello maniglia,

thanks for the quick reply. In my opinion the SetDisplayCallbacks is useful when debug, normally is not needed.
It helped me to see that the CubeProgrammerAPI.dll is working as expected over the USB and I was able to see were it got stuck when I've made a mistake. It seems to decrease the  CubeProgrammerAPI.dll speed when is used. I have tested the 32 bit variant without calling this function and it worked. The API documentation and the c++ examples are recommending to call this first along with the much critical SetLoadersPath. In your Programmer.dll SetDisplayCallbacks is called in dllmain.cpp at the DLL call.
Kind Regards,

**Reply by maniglia** [reply]
Well, then it will be time to introduce the possibility of enabling _dubug mode_.
I agree that it can be useful when you need to debug.

**Comment by maniglia** [comment]
There is a new version, the _0.1.70_.
I removed all references of the _SetDisplayCallbacks_ function from main in _Programmer.dll_, now you can call the _SetDisplayCallbacks_ 
 method.
I did some testing using the following example:

```
var displayCallBacks = new DisplayCallBacks
{
    InitProgressBar = null, 
    LogMessage = ReceiveMessage,
    LoadBar = null, 
};

CubeProgrammerApi.SetDisplayCallbacks(ref displayCallBacks);

CubeProgrammerApi.SetVerbosityLevel(CubeProgrammerVerbosityLevel.CubeprogrammerVerLevelDebug);
```

I'm having problems when I try to set the _InitProgressBar_ and _LoadBar_, I think it is necessary to test this feature thoroughly.
Any feedback is welcome.

**Comment by mihaisvcc** [comment]
Hello maniglia,

i share here how i use the SetDisplayCallbacks in the following way:

```
SetDisplayCallbacks(InitProgressBar, ReceiveMessage, ProgressBarUpdate);

  private void ReceiveMessage(int messgaeType, string message)
        {
            //only if I want to see message from the dll, more messages are coming when verbosity level is at 3
        
        }
        private void InitProgressBar()
        {

        }

        private void ProgressBarUpdate(int currentProgress, int total)
        {
            float current;
            if (total > 0)
            {
                //can use current variable to set a progress bar, I have noticed that currentProgress is advancing only by write or download 
                //operation, erase operation does not produce advance
                current = (float)((currentProgress * 100) / total);
                
            }
        }
```
My delegates

  ```
public delegate void LogMessageReceived(int messageType, [MarshalAs(UnmanagedType.LPWStr)] string message);
        public delegate void InitProgressBar();
        public delegate void ProgressBarUpdateReceived(int currentProgress, int total);

 public struct DisplayCallbacks
        {
            public InitProgressBar initProgressBar;
            public LogMessageReceived logMessage;
            public ProgressBarUpdateReceived loadBar;
        }

 ```
public static void SetDisplayCallbacks(InitProgressBar initProgressBar, LogMessageReceived messageReceived, ProgressBarUpdateReceived progressBarUpdate)
    {
        DisplayCallbacks callbacksHandle;

        callbacksHandle.initProgressBar = initProgressBar;
        callbacksHandle.logMessage = messageReceived;
        callbacksHandle.loadBar = progressBarUpdate;

        SetDisplayCallbacks(ref callbacksHandle); //call to dll import
    }

```
```

It works good under x64 under x32 the dll import crashes ...i don't know why ?

**Reply by maniglia** [reply]
It's strange I also tested 32 bit without problems, can you share a project with this problem?

**Comment by maniglia** [comment]
Hello @mihaisvcc,
I'm not sure I solved the problem, I think I forgot some decorators on the delegates in _DisplayCallBacks_ structure.
This is the content of the new version of the _DisplayCallBacks.cs_ file:

```
namespace KSociety.SharpCubeProgrammer.Struct
{
    using System.Runtime.InteropServices;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl, SetLastError = true)]
    public delegate void InitProgressBar();

    [UnmanagedFunctionPointer(CallingConvention.Cdecl, SetLastError = true)]
    public delegate void LogMessageReceived(int messageType, [MarshalAs(UnmanagedType.LPWStr)] string message);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl, SetLastError = true)]
    public delegate void ProgressBarUpdateReceived(int currentProgress, int total);

    /// <summary>
    /// Functions must be implemented to personalize the display of messages.
    /// </summary>
    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
    public struct DisplayCallBacks
    {
        public InitProgressBar? InitProgressBar;
        public LogMessageReceived? LogMessage;
        public ProgressBarUpdateReceived? LoadBar;
    }
}
```
Please let me know if you can test this code snippet.
Regards

**Comment by mihaisvcc** [comment]
Hello @maniglia 

Thank you for your answer and for the code snippet.
I have tested this...the problem seems to remain.
I got an access violation for the x86 driver if I compile as 32bit.
I have attached a test project made in VS2022 to see if you can reproduce the problem.
The test project is trimmed to a very basic level.
Now because the project file was very large, please copy inside the bin\debug and release folder the st and the 
dll\x86
   \x64
st\Programmer\FlashLoader
                         \ExternalLoader


Regards,

[CubeProgrammerAPIDirectTest.zip](https://github.com/K-Society/KSociety.SharpCubeProgrammer/files/13901313/CubeProgrammerAPIDirectTest.zip)




System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

**Comment by maniglia** [comment]
Hello @mihaisvcc,
to fix the exception, please edit the _CubeProgrammerApi.cs_ file, for _Native.SetDisplayCallbacks_ remove **ref** keyword:
Before:
```
[DllImport(Dll64, CallingConvention = CallingConvention.Cdecl, EntryPoint = "setDisplayCallbacks")]
internal static extern void SetDisplayCallbacks64(ref DisplayCallbacks callbacks);
[DllImport(Dll32, CallingConvention = CallingConvention.Cdecl, EntryPoint = "setDisplayCallbacks")]
internal static extern void SetDisplayCallbacks32(ref DisplayCallbacks callbacks);
internal static void SetDisplayCallbacks(ref DisplayCallbacks callbacks)
{
    if (Environment.Is64BitProcess)
    {
        SetDisplayCallbacks64(ref callbacks);
    }
    else
    {
        SetDisplayCallbacks32(ref callbacks);
    }
}
```

After:

```
[DllImport(Dll64, CallingConvention = CallingConvention.Cdecl, EntryPoint = "setDisplayCallbacks")]
internal static extern void SetDisplayCallbacks64(DisplayCallbacks callbacks);
[DllImport(Dll32, CallingConvention = CallingConvention.Cdecl, EntryPoint = "setDisplayCallbacks")]
internal static extern void SetDisplayCallbacks32(DisplayCallbacks callbacks);
internal static void SetDisplayCallbacks(DisplayCallbacks callbacks)
{
    if (Environment.Is64BitProcess)
    {
        SetDisplayCallbacks64(callbacks);
    }
    else
    {
       SetDisplayCallbacks32(callbacks);
    }
}
```

Consequently the _SetDisplayCallbacks_ method also changes:
```
public static void SetDisplayCallbacks(InitProgressBar initProgressBar, LogMessageReceived messageReceived, ProgressBarUpdateReceived progressBarUpdate)
{
    Native.DisplayCallbacks callbacksHandle;

    callbacksHandle.initProgressBar = initProgressBar;
    callbacksHandle.logMessage = messageReceived;
    callbacksHandle.loadBar = progressBarUpdate;

    //if (Environment.Is64BitProcess)
    //{
        Native.SetDisplayCallbacks(callbacksHandle); // ?? x86 crash
   // }
}
```
Regards

**Comment by mihaisvcc** [comment]
Hello @maniglia 

Thank you very much for update. Unfortunately this does not seem to solve my problem. 
I sitll got the memory corrupt error when I try to call the DisplayCallbacks x86 bit dll when my app si compiled as 32bit.
I have made a test project, very simple in VS2022. Please test this when you have time and tell me what you think.
In the test project I have deleted the x64 dll's in order to keep it small size.

Kind regards, 
[CubeProgrammerAPIDirectTest.zip](https://github.com/K-Society/KSociety.SharpCubeProgrammer/files/13903844/CubeProgrammerAPIDirectTest.zip)

**Comment by maniglia** [comment]
It works for me with the modified file, you sent me the same project again.

**Comment by mihaisvcc** [comment]
Hello @maniglia 

**Thank you for the correction tip** !!, It works now for me also. Sorry for the double post...after hitting the comment button the content was first displayed only after a couple of minutes, then I made the mistake to post it again.

**Comment by kvasilak** [comment]
First Awesome job. Made my life way easier.
But the progress bar acts weird, current often goes bigger than total. Hard to use for a progress bar :D
During the erase it will show 300%
Then programming will go to 200%

So I just show the current value to show progress

**Comment by maniglia** [comment]
Hi @kvasilak 
Unfortunately I have no control on progress bar.
The _ProgressBarUpdate_ function has two parameters: _currentProgress_ and _total_
In my example I use the following formula: `var current = (currentProgress * 100F) / total;`
I never checked the return value.
As soon as I find some time I'll do some testing...
